svdtools/patch.py

```diff
@@ -923,22 +946,37 @@ def process_field(self, pname, fspec, field):
 
     def process_field_enum(self, pname, fspec, field, usage="read-write"):
         """Add an enumeratedValues given by field to all fspec in rtag."""
-        derived = None
+        replace_if_exists = False
+        if "_replace_enum" in field:
+            field = field["_replace_enum"]
+            replace_if_exists = True
+
+        derived, enum, enum_name, enum_usage = None, None, None, None
         for ftag in self.iter_fields(fspec):
             name = ftag.find("name").text
-            if derived is None:
+
+            if enum is None:
                 enum = make_enumerated_values(name, field, usage=usage)
                 enum_name = enum.find("name").text
                 enum_usage = enum.find("usage").text
-                for ev in ftag.iter("enumeratedValues"):
-                    ev_usage = ev.find("usage").text
-                    if ev_usage == enum_usage or ev_usage == "read-write":
+
+            for ev in ftag.iter("enumeratedValues"):
+                ev_usage_tag = ev.find("usage")
+                ev_usage = (
+                    ev_usage_tag.text if ev_usage_tag is not None else "read-write"
+                )
+                if ev_usage == enum_usage or ev_usage == "read-write":
+                    if replace_if_exists:
+                        ftag.remove(ev)
+                    else:
                         print(pname, fspec, field)
                         raise SvdPatchError(
                             "{}: field {} already has enumeratedValues for {}".format(
                                 pname, name, ev_usage
                             )
                         )
```

Comment:
This seems to be quite hard to fix in stm32-rs; for a lot of occurrences it's not obvious how to fix it as the duplicate definition seems to often come from the same line of the patch yaml getting applied in different ways with derived peripherals etc. I think ideally we'd still fix the stm32-rs patches to prevent duplicates, but I wonder if we can instead just keep the existing enum if not specifically requested to replace it, like this?

```suggestion
                        # If we weren't asked to replace an existing enum,
                        # just leave the current one in place.
                        return
```

Reply:
Another option might be to merge non-duplicate variants by default and ignore duplicate variants, allowing variants to be specified in two places and merged together. That would be quite useful as often you can define the common variants in one shared file and then define a few extra variants for a specific device elsewhere.

Reply:
I'm a bit hesitant about those changes you suggested because both are behavior that would be easy to accidentally opt-in to, without knowing.  IMO the default should always be to fail with noise; if a patch author wants merging or replacing behavior, this should be stated explicitly in the patchfile.

Otherwise, we'll end up with the same issue, just differently:  The reason we have a problem right now is that svdpatch didn't error out on this arguably wrong patch earlier so the patches were deemed correct even if the generated SVD now contains multiple enumeratedValues.  Say we now switch to the "silently drop the new enum" or "merge enums" behavior; I can see how the same thing might happen again where people don't check the patched SVD file carefully and miss wrong/additional enum variants.

---

OTOH, apart from fixing the patches in stm32-rs, I don't really see a solution right now :( I'll think about it ...

Reply:
In the meantime I'll start working on the fixes for stm32-rs.  I'll try and gauge how difficult it really is to work through them ...

Reply:
Thanks, I'd appreciate it! I made a start but pretty quickly hit some weird roadblocks. We could try and chat through some of them on stm32-rs:matrix.org if you want. I'll keep thinking if there's another way around it too.

Reply:
Ugh, so it looks like the largest part of errors in `stm32-rs` are actually not even problems in the patches but an oversight in the bugfix I added here:  When checking for duplicates, the code wouldn't take into account the _usage_ in the case it encountered a derived `<enumeratedValues>` and always assumed a collision.

A naive fix could look like this (tested against stm32-rs, works as intended):

```python
for ev in ftag.iter("enumeratedValues"):
    if len(ev) == 0:
        # This is a derived enumeratedValues so guess the usage by
        # the name ...
        derived_from = ev.attrib["derivedFrom"]
        if derived_from.endswith("W"):
            ev_usage = "write"
        elif derived_from.endswith("R"):
            ev_usage = "read"
        else:
            ev_usage = "read-write"
    else:
        ev_usage = ev.find("usage").text

    if ev_usage == enum_usage or ev_usage == "read-write":
        print(pname, fspec, field)
        raise SvdPatchError(
            "{}: field {} already has enumeratedValues for {}".format(
                pname, name, ev_usage
            )
        )
```

This is, of course, a bit of a hack; the proper solution would be to resolve the element containing the definition of the derived enum and look up the usage there.  Do you think I should spend the time to try and implement that?

Reply:
Aaah, that totally explains why I was going around in circles trying to fix it all. Thanks for figuring it out. If you can easily resolve the derived enum that would be better; I think the real resolution process is non-trivial but I expect that it should  work fine to just check all sibling fields of this register, along the lines of `self.rtag.findall(f"./fields/field/enumeratedValues/[name='{derived_from}']")` perhaps.

Reply:
Looks like that's all it took.  Thanks for the hint!  See commit e62268b53e95 ("Find usage from derived enumeratedValues in duplication check") for my implementation.  I tested it against `stm32-rs`: It works well and does not produce any changes in the patched SVD file.